File: src/main/java/org/jetbrains/space/jenkins/config/SpacePluginConfiguration.java
package org.jetbrains.space.jenkins.config;

import hudson.Extension;
import hudson.model.Describable;
import hudson.model.Descriptor;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Objects;
import java.util.stream.Collectors;

/**
 * Piece of Jenkins configuration containing the list of org-level SpaceCode connections.
 */
@Extension
public class SpacePluginConfiguration extends Descriptor<SpacePluginConfiguration> implements Describable<SpacePluginConfiguration> {

    private List<SpaceConnection> connections;

    public SpacePluginConfiguration() {
        super(SpacePluginConfiguration.class);
        load();
        if (connections == null)
            connections = new ArrayList<>();
    }

    public List<SpaceConnection> getConnections() {
        return Collections.unmodifiableList(connections);
    }

    public void addConnection(SpaceConnection connection) {
        connections = new ArrayList<>(
                connections.stream()
                        .filter(c ->
                                !Objects.equals(c.getClientId(), connection.getClientId()) &&
                                        !c.getId().equals(connection.getId()))
                        .collect(Collectors.toList())
        );
        connections.add(connection);
        save();
    }

    /**
     * Called when a job, workflow or multibranch project is renamed in Jenkins.
     * Reflects the name change in the nested collections of child connections.
     */
    public void onItemRenamed(String oldFullName, String newFullName) {
        connections.forEach(conn -> conn.onItemRenamed(oldFullName, newFullName));
        save();
    }

    /**
     * Called when a job or workflow is updated in Jenkins.
     * Cleans up unused project-level connections, SSH credentials and SpaceCode applications that the job doesn't use anymore.
     */
    public void onJobUpdated(String jobFullName, SpaceProjectConnectionJobProperty jobProperty) {
        connections.forEach(parentConnection -> {
            SpaceProjectConnection projectConnection = parentConnection.getProjectConnectionsByJob().get(jobFullName);
            if (projectConnection != null && !isTheSameProjectConnection(jobProperty, parentConnection, projectConnection)) {
                parentConnection.getProjectConnectionsByJob().remove(jobFullName);
                SpaceProjectConnectionKt.deleteProjectApplication(projectConnection, parentConnection);
                SpaceProjectConnectionKt.deleteSshKeyCredentials(projectConnection);
            }
        });
    }

    private static boolean isTheSameProjectConnection(
            SpaceProjectConnectionJobProperty jobProperty,
            SpaceConnection rootConnection,
            SpaceProjectConnection projectConnection
    ) {
        return jobProperty != null
                && rootConnection.getId().equals(jobProperty.getSpaceConnectionId())
                && projectConnection.getProjectKey().equals(jobProperty.getProjectKey());
    }

    /**
     * Called when a job, workflow or multibranch project is removed.
     * Cleans up the child project-level connections, ssh credentials for this job/project
     * and requests SpaceCode to uninstall the corresponding applications.
     */
    public void onItemDeleted(String fullName) {
        connections.forEach(conn -> conn.onItemDeleted(fullName));
        save();
    }

    /**
     * Removes the org-level SpaceCode connection by its identifier.
     * Cleans up all child connections, related SpaceCode applications and SSH keys
     *
     * @return true if the connection was successfully removed, false if it hasn't been found by id.
     */
    public boolean removeConnectionById(String id) {
        SpaceConnection connection = connections.stream().filter((conn) -> conn.getId().equals(id)).findFirst().orElse(null);
        if (connection == null)
            return false;

        SpaceConnectionKt.deleteApplication(connection);
        this.connections = connections.stream().filter((conn) -> conn != connection).collect(Collectors.toList());
        save();
        return true;
    }

    /**
     * Called when SpaceCode notifies Jenkins about the uninstall of an application that supports the integration.
     * Cleans up the corresponding connection entries on Jenkins side and removes unused SSH credentials
     */
    public void onSpaceAppUninstalled(String clientId) {
        connections = connections.stream()
                .filter((conn) -> !conn.getClientId().equals(clientId))
                .collect(Collectors.toList());
        connections.forEach(conn -> conn.onSpaceAppUninstalled(clientId));
        save();
    }

    @Override
    public Descriptor<SpacePluginConfiguration> getDescriptor() {
        return this;
    }
}

File: src/main/java/org/jetbrains/space/jenkins/config/SpacePluginConfiguration.kt
package org.jetbrains.space.jenkins.config

import hudson.ExtensionList
import jenkins.branch.MultiBranchProject
import org.jetbrains.space.jenkins.scm.SpaceSCMSource
import java.util.ArrayList

fun getOrgConnection(id: String): SpaceConnection? {
    return ExtensionList.lookupSingleton<SpacePluginConfiguration>(SpacePluginConfiguration::class.java)
        .connections.firstOrNull { it.id == id }
}

/**
 * Retrieves the SpaceCode project connection and SpaceCode URL for a given Jenkins job, workflow or multibranch project.
 * <br />
 * Org-level connection id and project name do not uniquely identify a project connection in the case of multibranch projects,
 * because a single project can have multiple branch sources pointing to different SpaceCode organizations and projects.
 * Therefore for multibranch projects, SpaceCode project key is provided in addition to org-level connection id.
 */
fun getProjectConnection(jenkinsItemFullName: String, spaceConnectionId: String, projectKey: String?): Pair<SpaceProjectConnection, String> {
    return (if (projectKey != null)
        getProjectConnectionForMultiBranchProject(jenkinsItemFullName, spaceConnectionId, projectKey)
    else
        getProjectConnectionForJob(jenkinsItemFullName, spaceConnectionId))
        ?: error("No SpaceCode connection found by specified id")
}

/**
 * Retrieves the SpaceCode project connection and SpaceCode URL for a given Jenkins job or workflow and org-level connection id.
 */
private fun getProjectConnectionForJob(jobFullName: String, spaceConnectionId: String): Pair<SpaceProjectConnection, String>? {
    val rootConnection = getOrgConnection(spaceConnectionId) ?: return null
    return rootConnection.projectConnectionsByJob?.get(jobFullName)?.let { it to rootConnection.baseUrl }
}

/**
 * Retrieves the SpaceCode project connection and SpaceCode URL for a given Jenkins multibranch project, org-level connection id and SpaceCode project key.
 * <br />
 * Org-level connection id and project name do not uniquely identify a project connection in the case of multibranch projects,
 * because a single project can have multiple branch sources pointing to different SpaceCode organizations and projects.
 */
private fun getProjectConnectionForMultiBranchProject(projectFullName: String, spaceConnectionId: String, projectKey: String): Pair<SpaceProjectConnection, String>? {
    val rootConnection = getOrgConnection(spaceConnectionId) ?: return null
    return rootConnection.projectConnectionsByMultibranchFolder?.get(projectFullName)
        ?.let { it.firstOrNull { it.projectKey == projectKey } }
        ?.let { it to rootConnection.baseUrl }
}

fun getSpaceApiClientForMultiBranchProject(projectFullName: String, spaceConnectionId: String, projectKey: String) =
    getProjectConnectionForMultiBranchProject(projectFullName, spaceConnectionId, projectKey)
        ?.let { (projectConnection, spaceUrl) -> projectConnection.getApiClient(spaceUrl) }

/**
 * Called when a multibranch project is updated in Jenkins.
 * Cleans up unused project-level connections, SSH credentials and SpaceCode applications that aren't used by the project's branch sources anymore.
 */
fun SpacePluginConfiguration.onMultiBranchProjectUpdated(project: MultiBranchProject<*,*>) {
    val scmSources = project.getSCMSources().filterIsInstance<SpaceSCMSource>()
    if (scmSources.isEmpty()) {
        onItemDeleted(project.getFullName())
    } else {
        connections.forEach { parentConn ->
            parentConn.projectConnectionsByMultibranchFolder?.put(
                project.fullName,
                parentConn.projectConnectionsByMultibranchFolder.get(project.fullName)
                    ?.filter { projectConn ->
                        scmSources.any { it.spaceConnectionId == parentConn.id && it.projectKey == projectConn.projectKey }
                            .also {
                                if (!it) {
                                    projectConn.deleteProjectApplication(parentConn)
                                    projectConn.deleteSshKeyCredentials()
                                }
                            }
                    }
                    .let { ArrayList(it.orEmpty()) }
            )
        }
    }
}